source/tutorials/packaging/contributing-to-nixpkgs.md

```diff
@@ -0,0 +1,222 @@
+---
+myst:
+  html_meta:
+    "description lang=en": "Contributing to Nixpkgs"
+    "keywords": "Nix, Nixpkgs, contributing"
+---
+
+
+(contributing-to-nixpkgs)=
+
+# Contributing to Nixpkgs
+
+[Nixpkgs](https://github.com/nixos/nixpkgs) is one of the largest public repositories by contributions and *the* largest single collection of software packages available on the web.
+
+In this tutorial, which follows {ref}`the tutorial on packaging existing software <packaging-existing-software>`, you will learn how to prepare newly-packaged software for submission to Nixpkgs, becoming a Nixpkgs contributor in the process.
+
+## Tags, Revisions, and Releases
+An earlier version of the `icat.nix` definition created in the {ref}`Packaging Existing Software <packaging-existing-software>` tutorial used the `master` revision of the [upstream `icat` repository](https://github.com/atextor/icat), which would be suitable for individual use but isn't good practice for submission to Nixpkgs.
+
+It is much better to use a fixed revision corresponding to a particular release version of the software so maintainers (perhaps you!) can easily check whether the package should be updated.
+
+This was already done in the previous tutorial, but as a reminder, here are the steps:
+1. Check the upstream repository for tags or releases. On GitHub, these are available at `https://github.com/<owner>/<repo>/tags` and `https://github.com/<owner>/<repo>/releases` respectively.
+2. Choose the latest tag or release and write it into a string literal as the value of the `rev` attribute passed to the [fetcher](https://nixos.org/manual/nixpkgs/stable/#chap-pkgs-fetchers).
+3. Use `nix-prefetch-url` or a similar tool to fetch, unpack, and hash the files marked by the tag or release you selected.
+4. Alternatively, you can also use a tool such as `nix-init` to automatically generate a required derivation for some common project types, or use its output to help lay down a skeleton implementation of the above steps.
+
+In the `icat` example, the source was hosted on GitHub, so `fetchFromGitHub` was used for the `src` attribute passed to `stdenv.mkDerivation`, with `rev = "v0.5"` corresponding to the latest tag, and a SHA256 hash generated by
+
+```console
+nix-prefetch-url --unpack https://github.com/atextor/icat/archive/refs/tags/v0.5.tar.gz --type sha256`
+```
+
+Here is the final version of `icat.nix` from the previous tutorial:
+
+```nix
+# icat.nix
+{ lib
+, stdenv
+, fetchFromGitHub
+, imlib2
+, xorg
+}:
+
+stdenv.mkDerivation {
+  pname = "icat";
+  version = "0.5";
+
+  src = fetchFromGitHub {
+    owner = "atextor";
+    repo = "icat";
+    rev = "v${version}";
+    sha256 = "0wyy2ksxp95vnh71ybj1bbmqd5ggp13x3mk37pzr99ljs9awy8ka";
+  };
+
+  buildInputs = [ imlib2 xorg.libX11.dev ];
+
+  installPhase = ''
+    runHook preInstall
+    mkdir -p $out/bin
+    cp icat $out/bin
+    runHook postInstall
+  '';
+}
```

Comment:
```suggestion
stdenv.mkDerivation (finalAttrs: {
  pname = "icat";
  version = "0.5";

  src = fetchFromGitHub {
    owner = "atextor";
    repo = "icat";
    rev = "v${finalAttrs.version}";
    sha256 = "0wyy2ksxp95vnh71ybj1bbmqd5ggp13x3mk37pzr99ljs9awy8ka";
  };

  buildInputs = [ imlib2 xorg.libX11.dev ];

  installPhase = ''
    runHook preInstall
    mkdir -p $out/bin
    cp icat $out/bin
    runHook postInstall
  '';
})
```

The `version` variable was undefined.

Reply:
@infinisil Should we just give it some placeholder like `0.1` then? In order to keep the focus on the content rather than on how `mkDerivation` is implemented? (I'd need to chew on this a bit myself before I can totally understand why your change works. I believe it has to do with how `mkDerivation` is defined in the first place, [in that it can take both a function or attribute set as an argument](https://github.com/NixOS/nixpkgs/blob/3a16c6447466f4034c2d75fe7014477142c9513e/pkgs/stdenv/generic/make-derivation.nix#L600), so that it is ultimately something like a fixed point function? Totally new to me, as I never dove into the source code on this point until trying to figure out why your change would work.)